File: python/lzss.py
'''LZSS compression'''

import sys

MAX_DISTANCE = 4095 + 1
MAX_LENGTH = 15 + 3

def _link_matches(data):
    offsets = [0] * len(data)
    prev = [-1] * (256 * 256)
    for i in range(len(data) - 1):
        h = data[i] * 256 + data[i + 1]
        offsets[i] = i - prev[h] if prev[h] >= 0 else 0
        prev[h] = i
    return offsets
    
def _best_match(data, offsets, j):
    if offsets[j] == 0: return (0, 0)
    best = (0, 0)
    ndata = len(data)
    i = j - offsets[j]
    while j - i <= MAX_DISTANCE:
        n = 0
        while n < MAX_LENGTH and j + n < ndata and data[i + n] == data[j + n]: n += 1
        if n > best[1]:
            best = (j - i, n)
            if best[1] == MAX_LENGTH: break
        if offsets[i] == 0: break
        i -= offsets[i]
    return best

def compress(data):
    offsets = _link_matches(data)
    compressed = bytearray()
    blk = bytearray(1)
    blk_n = 0
    j = 0
    ndata = len(data)
    while j < ndata:
        (d, n) = _best_match(data, offsets, j)
        if n > 2:
            blk.append((n - 3) * 16 + (d - 1) // 256)
            blk.append((d - 1) % 256)
            j += n
        else:
            blk[0] |= 0x80 >> blk_n
            blk.append(data[j])
            j += 1
        blk_n += 1
        if blk_n == 8:
            compressed.extend(blk)
            blk = bytearray(1)
            blk_n = 0
    if blk_n > 0: compressed.extend(blk)
    return bytes(compressed)
    
if __name__ == '__main__':
    if len(sys.argv) < 3:
        print('use: lzss <source-file> <target-file>')
        exit(1)
    elif sys.argv[1] == '-e' and len(sys.argv) < 4:
        print('use: lzss <source-file> <target-file>')
        exit(1)
    
    if sys.argv[1] == '-e':
        pass
    else:
        with open(sys.argv[1], 'rb') as input:
            compressed = compress(input.read())
            with open(sys.argv[2], 'wb') as output:
                output.write(compressed)

File: kotlin/lzss.kt
// LZSS compression

import java.io.ByteArrayOutputStream
import java.io.File
import kotlin.system.exitProcess

const val MAX_DISTANCE = 4095 + 1
const val MAX_LENGTH = 15 + 3

fun linkMatches(data : ByteArray) : IntArray {
    val prev = IntArray(256 * 256) { -1 }
    val offsets = IntArray(data.size) { 0 }
    for (i in 0 until data.size - 1) {
        val h = (data[i].toInt() and 0xFF) * 256 + (data[i + 1].toInt() and 0xFF)
        offsets[i] = if (prev[h] < 0) 0 else i - prev[h]
        prev[h] = i
    }
    return offsets;
}

fun bestMatch(data: ByteArray, offsets: IntArray, j: Int): Pair<Int, Int> {
    var best = Pair(0, 0)
    if (offsets[j] == 0) return best

    var i = j - offsets[j]
    while (j - i <= MAX_DISTANCE) {
        var n = 2
        while (j + n < data.size && data[i + n] == data[j + n] && n < MAX_LENGTH) n += 1
        if (n > best.second) best = Pair(j - i, n)
        if (offsets[i] == 0) break
        i -= offsets[i]
    }

    return best
}

fun compress(data: ByteArray): ByteArray {
    val compressed = ByteArrayOutputStream()

    var blk = ByteArray(17)
    var blkI = 1
    var blkN = 0

    val offsets = linkMatches(data)
    var j = 0
    while (j < data.size) {
        val (d, n) = bestMatch(data, offsets, j)
        if (n > 2) {
            blk[blkI++] = (((n - 3) shl 4) + (d - 1) / 256).toByte()
            blk[blkI++] = ((d - 1) % 256).toByte()
            j += n
        }
        else {
            blk[0] = (blk[0].toInt() or (0x80 shr blkN)).toByte()
            blk[blkI++] = data[j++]
        }

        if (++blkN == 8) {
            compressed.write(blk, 0, blkI)
            blk[0] = 0
            blkI = 1
            blkN = 0
        }
    }

    if (blkN > 0) compressed.write(blk, 0, blkI)
    return compressed.toByteArray()
}

fun expand(data: ByteArray): ByteArray {
    var expanded = ByteArray(65536)
    var expandedN = 0

    var blk = 0
    var blkN = 8

    var j = 0
    while (j < data.size) {
        if (++blkN >= 8) {
            blk = data[j++].toInt() and 0xFF
            blkN = 0
        }

        if ((blk and (0x80 shr blkN)) == 0) {
            var n = (data[j].toInt() and 0xF0) / 16 + 3
            val d = (data[j].toInt() and 0x0F) * 256 + (data[j + 1].toInt() and 0xFF) + 1
            if (expandedN + n > expanded.size)
                expanded = expanded.copyOf(expanded.size + expanded.size / 2)
            while (n-- > 0) {
                expanded[expandedN] = expanded[expandedN - d]
                expandedN ++
            }
            j += 2
        }
        else {
            if (expanded.size == expandedN)
                expanded = expanded.copyOf(expanded.size + expanded.size / 2)
            expanded[expandedN++] = data[j++]
        }
    }

    return expanded.copyOf(expandedN)
}

fun readFile(name: String): ByteArray {
    return File(name).readBytes()
}

fun writeFile(name: String, data: ByteArray) {
    File(name).writeBytes(data)
}

fun main(args: Array<String>) {
    if (args.size < 2) {
        println("use: lzss <source-file> <target-file>")
        exitProcess(1)
    }
    else if (args[0] == "-e" && args.size < 3) {
        println("use: lzss <source-file> <target-file>")
        exitProcess(1)
    }

    if (args[0] == "-e") {
        writeFile(args[2], expand(readFile(args[1])))
    }
    else {
        writeFile(args[1], compress(readFile(args[0])))
    }
}